### PR TITLE
Add speed/effort parameter

### PR DESCRIPTION
**Pull request by snowfluke**
I believe this parameter is necessary for balancing between speed and quality.

Parameter references: https://github.com/strukturag/libheif/blob/2dff6b4282d5e03feb5bd8dea5b09710e635cf68/libheif/heif_encoder_aom.cc#L70

### CODE REVIEW

**Comment by snowfluke** [comment]
or maybe add a higher level API to set a custom parameter would be great

**Comment by awxkee** [comment]
Before I merge I have some suggestion.

The most of users do not really know anything about speed/effort and how it works. Especially for AV1 where each encoder provides it own implementation of effors/speed.

So, my suggestion to introduce small brief description how it works, what direction will increase/decrease an effect and might be good to introduce speed parameter as enum in kotlin with readable names or otherwise protect the parameter with `@IntRange(0, 20)`.

File: avif-coder/src/main/java/com/radzivon/bartoshyk/avif/coder/HeifCoder.kt [hunk 1]
```diff
@@ -116,25 +116,25 @@ class HeifCoder(
         )
     }
 
-    fun encodeAvif(bitmap: Bitmap, quality: Int = 80, preciseMode: PreciseMode = PreciseMode.LOSSY): ByteArray {
+    fun encodeAvif(bitmap: Bitmap, quality: Int = 80, speed: Int = 0, preciseMode: PreciseMode = PreciseMode.LOSSY): ByteArray {
```
Comment (awxkee) [inline avif-coder/src/main/java/com/radzivon/bartoshyk/avif/coder/HeifCoder.kt:119]: If you'lll decode not do enum, protect with `@IntRange` will be good

File: avif-coder/src/main/java/com/radzivon/bartoshyk/avif/coder/HeifCoder.kt [hunk 2]
```diff
@@ -116,25 +116,25 @@ class HeifCoder(
...
-            encodeAvifImpl(bitmap, quality, bitmap.colorSpace?.dataSpace ?: -1, preciseMode.value)
+            encodeAvifImpl(bitmap, quality, speed, bitmap.colorSpace?.dataSpace ?: -1, preciseMode.value)
         } else {
-            encodeAvifImpl(bitmap, quality, -1, preciseMode.value)
+            encodeAvifImpl(bitmap, quality, speed, -1, preciseMode.value)
         }
     }
 
-    fun encodeHeic(bitmap: Bitmap, quality: Int = 80, preciseMode: PreciseMode = PreciseMode.LOSSY): ByteArray {
+    fun encodeHeic(bitmap: Bitmap, quality: Int = 80, speed: Int = 0, preciseMode: PreciseMode = PreciseMode.LOSSY): ByteArray {
```
Comment (awxkee) [inline avif-coder/src/main/java/com/radzivon/bartoshyk/avif/coder/HeifCoder.kt:130]: If you'lll decode not do enum, protect with `@IntRange` will be good

**Review by awxkee — CHANGES_REQUESTED** [review]

File: avif-coder/src/main/java/com/radzivon/bartoshyk/avif/coder/AvifSpeed.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+package com.radzivon.bartoshyk.avif.coder
+
+/**
+ * Enum representing speed values from 0 to 10 (slowest-fastest).
+ * Default iis 6.
```
[the remark this answers is not in the document]
Comment (snowfluke) [reply avif-coder/src/main/java/com/radzivon/bartoshyk/avif/coder/AvifSpeed.kt:33]: gotcha

File: avif-coder/src/main/cpp/JniEncoder.cpp [hunk 4]
```diff
@@ -114,6 +114,15 @@ jbyteArray encodeBitmap(JNIEnv *env, jobject thiz,
         throwException(env, str);
         return static_cast<jbyteArray>(nullptr);
       }
+      if(speed > 0 && speed < 10){
```
[the remark this answers is not in the document]
Comment (snowfluke) [reply avif-coder/src/main/cpp/JniEncoder.cpp:117]: Couldn't agree more, gotta fix it real quick. Sorry for troubling you, I'm not used in contributing

[the remark this answers is not in the document]
Comment (awxkee) [reply avif-coder/src/main/cpp/JniEncoder.cpp:117]: No worries:)

**Comment by snowfluke** [comment]
I add new enum value and fix the checking, in case of using `x265` instead of `aom` I don't know much about the implementation of it

**Comment by awxkee** [comment]
> I add new enum value and fix the checking, in case of using `x265` instead of `aom` I don't know much about the implementation of it

Actually you don't have to expose -1 to kotlin, you may pass it just in c++ to avoid suspicious enum values. However that's ok.  I think I'll do that changes myself it'll be easier

**Pull request merged**